src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,694 @@
+    size_t primary_key_pos = getPrimaryKeyPos(sample_block, getPrimaryKey());
+
+    if (null_map)
+    {
+        null_map->clear();
+        null_map->resize_fill(keys.size(), 1);
+    }
+
+    auto client = getClient();
+
+    std::vector<std::future<Coordination::GetResponse>> values;
+    values.reserve(keys.size());
+
+    for (const auto & key : keys)
+    {
+        const auto full_path = fullPathForKey(key);
+        if (full_path == metadata_path)
+        {
+            values.emplace_back();
+            continue;
+        }
+
+        values.emplace_back(client->asyncTryGet(full_path));
+    }
+
+    auto wait_until = std::chrono::system_clock::now() + std::chrono::milliseconds(Coordination::DEFAULT_OPERATION_TIMEOUT_MS);
+
+    for (size_t i = 0; i < keys.size(); ++i)
+    {
+        auto & value = values[i];
+        if (!value.valid())
+            continue;
+
+        if (value.wait_until(wait_until) != std::future_status::ready)
+            throw DB::Exception(ErrorCodes::KEEPER_EXCEPTION, "Failed to fetch values: timeout");
+
+        auto response = value.get();
+        Coordination::Error code = response.error;
+
+        if (code == Coordination::Error::ZOK)
+        {
+            fillColumns(base64Decode(keys[i], true), response.data, primary_key_pos, sample_block, columns);
+        }
+        else if (code == Coordination::Error::ZNONODE)
+        {
+            if (null_map)
+            {
+                (*null_map)[i] = 0;
+                for (size_t col_idx = 0; col_idx < sample_block.columns(); ++col_idx)
+                    columns[col_idx]->insert(sample_block.getByPosition(col_idx).type->getDefault());
+            }
+        }
+        else
+        {
+            throw DB::Exception(ErrorCodes::KEEPER_EXCEPTION, "Failed to fetch value: {}", code);
+        }
+    }
+
+    size_t num_rows = columns.at(0)->size();
+    return Chunk(std::move(columns), num_rows);
+}
+
+Block StorageKeeperMap::getSampleBlock(const Names &) const
+{
+    auto metadata = getInMemoryMetadataPtr();
+    return metadata ? metadata->getSampleBlock() : Block();
+}
+
+namespace
+{
+
+StoragePtr create(const StorageFactory::Arguments & args)
+{
+    ASTs & engine_args = args.engine_args;
+    if (engine_args.empty() || engine_args.size() > 3)
+        throw Exception(
+            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+            "Storage KeeperMap requires 1-3 arguments:\n"
+            "root_path: path in the Keeper where the values will be stored (required)\n"
+            "create_missing_root_path: 1 if the root path should be created if it's missing, otherwise throw exception (default: 1)\n",
+            "keys_limit: number of keys allowed to be stored, 0 is no limit (default: 0)");
+
+    const auto root_path_node = evaluateConstantExpressionAsLiteral(engine_args[0], args.getLocalContext());
+    auto root_path = checkAndGetLiteralArgument<std::string>(root_path_node, "root_path");
```

Comment:
Should we also do some macro substitutions? (see `Context::getMacros()` and `Macros.h`)

Reply:
Is this required? If not, I would like to create an issue and do it in a different PR.